### ISSUE TITLE
Using new youtube video chapters

### ISSUE THREAD

**Issue by daviddavo** [web-scrobbler/web-scrobbler#2509]
**Is your feature request related to a problem? Please describe.**
With new Youtube's feature where a video is divided in "parts", lots of people started dividing one 58 minute long video (like a music album or a playlist) into "parts" that you can click. Here's an [example](https://www.youtube.com/watch?v=6GbD20h8-_4).

Here is their [specification](https://support.google.com/youtube/answer/9884579?hl=en)

**Describe the solution you'd like**
It should be as easy as reading the video description and scrobbling according to which part you are listening to at the moment

**Additional context**
The format of the description usually is something like:

```
Enigma Variations, Variations on an Original Theme for Orchestra Op. 36, "Enigma"‏, Sir Edward Elgar, 1898-1899:
00:00 Theme (Andante)
01:56 Variation I. (L'istesso tempo) "C.A.E."
04:09 Variation II. (Allegro) "H.D.S.- P."
04:56 Variation III. (Allegretto) "R.B.T."
06:19 Variation IV. (Allegro di molto) "W.M.B."
06:49 Variation V. (Moderato) "R.P.A."
09:07 Variation VI. (Andantino) "Ysobel"
10:27 Variation VII. (Presto) "Troyte"
11:22 Variation VIII. (Allegretto) "W.N."
13:29 Variation IX. (Adagio) "Nimrod" -  http://youtu.be/iPlW4pAFMOM
19:05 Variation X. (Intermezzo : Allegretto) "Dorabella"
22:14 Variation XI. (Allegro di molto) "G.R.S."
23:22 Variation XII. (Andante) "B.G.N."
26:45 Variation XIII. (Romanza: Moderato) "* * *"
29:51 Variation XIV. (Finale: Allegro - Presto) "E.D.U."
BBC Symphony Orchestra, Leonard Bernstein.
Live Broadcast, 1984
```

**Comment by alexesprit** [comment]
Closed as a duplicate of #2371.

FYI: it's already implemented in d953547b15a013c10fa12cb6b4dfaf6aaeffbde6.

**Comment by jaccarmac** [comment]
It appears that the current implementation of chapters works for mixes where each chapter has an artist and title, but not in cases where for example an album is uploaded and the chapters are song names. I'm not running the tip of master though, so I'll wait on that to confirm.

**Comment by alexesprit** [comment]
> It appears that the current implementation of chapters works for mixes where each chapter has an artist and title, but not in cases where for example an album is uploaded and the chapters are song names.

Yes, you're right. It's quite hard to parse album videos. :(

**Comment by jaccarmac** [comment]
I whipped up this new implementation of `getTrackInfoFromChapters` and it seems to work broadly speaking, while still working on the mix mentioned in this video. Is there some refining I can do to get a similar patch into an acceptable state?

```js
let { artist, track } = Util.processYtVideoTitle(Util.getTextFromSelectors(chapterNameSelector));
if (!artist) {
	({ artist } = Util.processYtVideoTitle(Util.getTextFromSelectors(videoTitleSelector)));
}
if (!artist) {
	artist = Util.getTextFromSelectors(channelNameSelector);
}
return { artist, track };
```

**Comment by alexesprit** [comment]
Unfortunately, I don' use YouTube for scrobbling music, and I barely understand, how it should work properly for chapters (as there're a lot of edge cases, and no unified format for music tracks).

I think this implementation would work. If it doesn't break the current behavior, I'm for this way.

> Is there some refining I can do to get a similar patch into an acceptable state?

It gets more complicated each time. I would add some simple caching/memoization, like I did for parsing a video description, but it could be enabled globally for the whole `Connector.getTrackInfo` method.

**Comment by jaccarmac** [comment]
Sounds good. The fallthrough is correct and results in more correct behavior, but there are some obvious failure cases that I can address up front. I'll refine the patch a little and then PR.
